presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java

```diff
@@ -261,13 +261,15 @@ public PlanNode visitExchange(ExchangeNode node, RewriteContext<Void> context)
                     .collect(toImmutableList());
 
             List<List<Symbol>> inputs = new ArrayList<>();
+            Map<List<Symbol>, Symbol> inputsToOutputs = new HashMap<>();
             for (int i = 0; i < node.getInputs().size(); i++) {
                 inputs.add(new ArrayList<>());
             }
             Set<Symbol> addedOutputs = new HashSet<>();
             ImmutableList.Builder<Symbol> outputs = ImmutableList.builder();
             for (int symbolIndex = 0; symbolIndex < node.getOutputSymbols().size(); symbolIndex++) {
-                Symbol canonicalOutput = canonicalize(node.getOutputSymbols().get(symbolIndex));
+                Symbol canonicalOutput = canonicalizeExchangeNodeOutput(node, symbolIndex, inputsToOutputs);
```

Comment:
This is confusing and hard to follow. This method processes each output, but on each loop, it modifies the inputsToOutputs map that is used by the next loop. It's hard to reason about the side effects.